OsmAnd/src/net/osmand/plus/myplaces/tracks/filters/RangeTrackFilter.kt

```diff
@@ -37,69 +39,244 @@ abstract class RangeTrackFilter(
 	}
 
 	open fun getDisplayMinValue(): Int {
-		return floor(minValue).toInt()
+		val formattedValue = getFormattedValue(flor(minValue))
+		return formattedValue.valueSrc.toInt()
+	}
+
+	private fun flor(value: T): String {
+		return when (value) {
+			is Float -> {
+				floor(value as Float).toString()
+			}
+
+			is Double -> {
+				floor(value as Double).toString()
+			}
+
+			else -> {
+				value.toString()
+			}
+		}
+	}
+
+	private fun ceil(value: T): String {
+		return when (value) {
+			is Float -> {
+				ceil(value as Float).toString()
+			}
+
+			is Double -> {
+				ceil(value as Double).toString()
+			}
+
+			else -> {
+				value.toString()
+			}
+		}
+	}
+
+	private fun getDisplayValue(value: T): Int {
+		val formattedValue = getFormattedValue(ceil(maxValue))
+		return formattedValue.valueSrc.toInt()
 	}
 
 	open fun getDisplayMaxValue(): Int {
-		return ceil(maxValue).toInt()
+		val formattedValue = getFormattedValue(ceil(maxValue))
+		return formattedValue.valueSrc.toInt() + 1
 	}
 
 	open fun getDisplayValueFrom(): Int {
-		return floor(valueFrom).toInt()
+		val formattedValue = getFormattedValue(valueFrom.toString())
+		return formattedValue.valueSrc.toInt()
 	}
 
 	open fun getDisplayValueTo(): Int {
-		return ceil(valueTo).toInt()
+		val formattedValue = getFormattedValue(valueTo.toString())
+		return formattedValue.valueSrc.toInt()
+	}
+
+	fun getBaseValueFromFormatted(value: String): Float {
+		val metricsConstants: MetricsConstants = app.settings.METRIC_SYSTEM.get()
+		val mode = app.settings.applicationMode
+		val speedConstant = app.settings.SPEED_SYSTEM.getModeValue(mode)
+		return when (filterType.measureUnitType) {
+			MeasureUnitType.SPEED -> OsmAndFormatter.getBaseValueFromFormattedSpeed(
+				value.toFloat(),
+				app,
+				speedConstant)
+
+			MeasureUnitType.ALTITUDE -> OsmAndFormatter.getMetersFromFormattedAltitudeValue(
+				value.toFloat(),
+				metricsConstants)
+
+			MeasureUnitType.DISTANCE -> OsmAndFormatter.getMetersFromFormattedDistanceValue(
+				value.toFloat(),
+				metricsConstants)
+
+			else -> value.toFloat()
+		}
+
 	}
 
+	private fun getFormattedValue(value: String): FormattedValue {
+		val metricsConstants: MetricsConstants = app.settings.METRIC_SYSTEM.get()
+		return when (filterType.measureUnitType) {
+			MeasureUnitType.SPEED -> OsmAndFormatter.getFormattedSpeedValue(value.toFloat(), app)
+			MeasureUnitType.ALTITUDE -> OsmAndFormatter.getFormattedAltitudeValue(
+				value.toDouble(),
+				app,
+				metricsConstants)
+
+			MeasureUnitType.DISTANCE -> OsmAndFormatter.getFormattedDistanceValue(
+				value.toFloat(),
+				app,
+				true,
+				metricsConstants)
+
+			else -> FormattedValue(value.toFloat(), value.toString(), "")
+		}
+	}
+
+	private fun getProperty(): GpxParameter {
+		return filterType.propertyList[0]
+	}
+
+	private fun getComparableValue(value: Any): T {
+		if (value is Number) {
+			return if (getProperty().typeClass == java.lang.Integer::class.java) {
```

Comment:
Looks strange and hard to support (if we add new number types, we must fix this method), but I am not sure how we could improve it

Reply:
now it covers most of possible types of data. extending list of supported types shouldn't be very often